runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java

```diff
@@ -136,141 +130,150 @@ public int compare(PackageAttributes o1, PackageAttributes o2) {
-            @Override
-            public PackageAttributes call() throws Exception {
-              return createPackageAttributes(file, stagingPath, packageName);
+            PackageAttributes attributes = PackageAttributes.forFileToStage(file, stagingPath);
+            if (source.getName() != null) {
+              attributes = attributes.withPackageName(source.getName());
             }
-          });
-      futures.add(future);
-    }
+            return attributes;
+          }
+        });
+  }
 
+  private boolean alreadyStaged(PackageAttributes attributes) throws IOException {
     try {
-      return Futures.allAsList(futures).get();
-    } catch (InterruptedException e) {
-      Thread.currentThread().interrupt();
-      throw new RuntimeException("Interrupted while staging packages", e);
-    } catch (ExecutionException e) {
-      throw new RuntimeException("Error while staging packages", e.getCause());
+      long remoteLength =
+          FileSystems.matchSingleFileSpec(attributes.getDestination().getLocation()).sizeBytes();
+      return remoteLength == attributes.getSize();
+    } catch (FileNotFoundException expected) {
+      // If the file doesn't exist, it means we need to upload it.
+      return false;
     }
   }
 
-  private static WritableByteChannel makeWriter(String target, CreateOptions createOptions)
-      throws IOException {
-    return FileSystems.create(FileSystems.matchNewResource(target, false), createOptions);
+  /** Stages one file ("package") if necessary. */
+  public ListenableFuture<StagingResult> stagePackage(
+      final PackageAttributes attributes,
+      final Sleeper retrySleeper,
+      final CreateOptions createOptions) {
+    return executorService.submit(
+        new Callable<StagingResult>() {
+          @Override
+          public StagingResult call() throws Exception {
+            return stagePackageSynchronously(attributes, retrySleeper, createOptions);
+          }
+        });
   }
 
-  /**
-   * Utility to verify whether a package has already been staged and, if not, copy it to the
-   * staging location.
-   */
-  private static void stageOnePackage(
-      PackageAttributes attributes, AtomicInteger numUploaded, AtomicInteger numCached,
-      Sleeper retrySleeper, CreateOptions createOptions) {
-    String source = attributes.getSourcePath();
-    String target = attributes.getDataflowPackage().getLocation();
+  private StagingResult stagePackageSynchronously(
+      PackageAttributes attributes, Sleeper retrySleeper, CreateOptions createOptions) {
+    try {
+      return stagePackageBlithely(attributes, retrySleeper, createOptions);
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format(
+              "Could not stage classpath element: %s to %s",
+              attributes.getSource(), attributes.getDestination().getLocation()),
+          e);
+    }
+  }
+
+  private StagingResult stagePackageBlithely(
```

Comment:
<!--thread_id:cc_142239914_t; commit:2eb67949d13a8f290b7d29c2898b2c1ed2dcf780; resolved:1-->
<!--section:context-quote-->
> **kennknowles** wrote:
> "Crashily"?

<!--section:body-->
Done. Split into tryStagePackage which is trivial straight-line code, tryStagePackageWithRetries that does just that loop, crashing if it hits something unrecoverable, and stagePackageSynchronously that just decorates the error message.